### PR TITLE
DNN-10072 Fix assembly version check

### PR DESCRIPTION
**Pull request by bdukes**
When an assembly already exists for a package, it's checked against a random
version to determine whether it's newer or older, rather than the latest
version

[DNN-10072](https://dnntracker.atlassian.net/browse/DNN-10072)
>When upgrading a package (i.e. installing a new version of an existing package), or doing a repair install, it is possible for an assembly in that package to overwrite a newer version of an existing assembly (e.g an assembly which was installed via a different package).
># Steps to Reproduce
>
> -    Install [PackageOne_4.1.0.zip](https://dnntracker.atlassian.net/secure/attachment/55238/55238_PackageOne_4.1.0.zip) (which references HtmlTags.dll version 4.1.0)
> -    Install [PackageTwo_4.2.0.zip](https://dnntracker.atlassian.net/secure/attachment/55237/55237_PackageTwo_4.2.0.zip) (which references HtmlTags.dll version 4.2.0)
> -    Put one of the modules on a page (both modules do the same thing), the message should say that version 4.2.0 is installed
> -    Re-install [PackageOne_4.1.0.zip](https://dnntracker.atlassian.net/secure/attachment/55238/55238_PackageOne_4.1.0.zip) (i.e. do a repair install)
>
> ## Expected Results
>
> -    HtmlTags 4.2.0 is still the referenced assembly
>
>## Actual Results
>
> -    HtmlTags 4.1.0 is the referenced assembly
>
>Note that this bug is non-deterministic, so it's possible that this scenario wouldn't trigger the bug, but I think it's probable that this will fall into the bug condition.

### CODE REVIEW

File: Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Schema.SqlDataProvider [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿/************************************************************/
+﻿regi/************************************************************/
```
Comment (galatrash-at-dnn) [inline Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Schema.SqlDataProvider]: Syntax error - ``regi`` must be removed

**Comment by bdukes** [comment]
Eek, sorry that crept in there. I've pushed the fix @galatrash-at-dnn

File: Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Schema.SqlDataProvider [hunk 2]
```diff
@@ -19496,7 +19496,7 @@ As
 	ELSE
 		BEGIN
 			-- AssemblyID is not null - Assembly is registered - test for version
-			SET @CurrentVersion  = (SELECT Version 
+			SET @CurrentVersion  = (SELECT Max(Version )
```
Comment (galatrash-at-dnn) [inline Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Schema.SqlDataProvider]: There is an issue with this comparison. Consider this SQL script and see how it identifies the wrong version as maximum.
```
CREATE TABLE #versions (ver nvarchar(100));
INSERT INTO #versions (ver) VALUES (N'2.0.0');
INSERT INTO #versions (ver) VALUES (N'10.0.0');

select * from #versions;
select MAX(ver) from #versions; -- identifies 2.0.0 as the maximum and not 10.0.0

DROP TABLE #versions;
```

Comment (bdukes) [inline Website/Providers/DataProviders/SqlDataProvider/DotNetNuke.Schema.SqlDataProvider]: @galatrash-at-dnn I've pushed a change which properly uses a function to parse the version into integers and then sorts based on that.  Does that look better?

**Comment by bdukes** [comment]
@galatrash-at-dnn is there anything else you need from me on this PR?

**Comment by galatrash-at-dnn** [comment]
@bdukes, I was just waiting the PM to add the Jira issue in the current sprint to merge. I'll review it today.

File: Website/Providers/DataProviders/SqlDataProvider/09.02.00.01.SqlDataProvider [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+				@AssemblyName,
+				@Version
+			)
+		END
+	ELSE
+		BEGIN
+			-- AssemblyID is not null - Assembly is registered - test for version
+			SET @CurrentVersion  = (SELECT TOP 1 a.Version
+									FROM {databaseOwner}{objectQualifier}Assemblies a
+									CROSS APPLY {databaseOwner}[{objectQualifier}fn_ParseVersion](a.Version)
```
Comment (galatrash-at-dnn) [inline Website/Providers/DataProviders/SqlDataProvider/09.02.00.01.SqlDataProvider]: muas add an alias for this line as
```CROSS APPLY {databaseOwner}[{objectQualifier}fn_ParseVersion](a.Version) AS v```

File: Website/Providers/DataProviders/SqlDataProvider/09.02.00.01.SqlDataProvider [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+							WHERE PackageID = @PackageID
+								AND AssemblyName = @AssemblyName)
+
+	IF @AssemblyID IS NULL
+		BEGIN
+			-- AssemblyID is null (not registered)
+			-- but assembly may be registerd by other packages so check for Max unstalled version
+			SET @CurrentVersion  = (SELECT TOP 1 a.Version
+									FROM {databaseOwner}{objectQualifier}Assemblies a
+									CROSS APPLY {databaseOwner}[{objectQualifier}fn_ParseVersion](a.Version)
```
Comment (galatrash-at-dnn) [inline Website/Providers/DataProviders/SqlDataProvider/09.02.00.01.SqlDataProvider]: must add an alias for this line as
```CROSS APPLY {databaseOwner}[{objectQualifier}fn_ParseVersion](a.Version) AS v```

**Comment by galatrash-at-dnn** [comment]
@bdukes, any progress on this?

**Comment by bdukes** [comment]
@galatrash-at-dnn changes have been made, let me know if there's anything else.

**Comment by sleupold** [comment]
the solution with a temp table is performance wise not optimal - this is not significant for current use case, but if someone uses function fn_compareversion in a larger query, it might become an issue. I am using the following function in my scripts.
`CREATE FUNCTION {databaseOwner}[{objectQualifier}VersionNumberSerial]
(
    @VersionString  nVarChar(20) -- String containing a version format '01.02.03'
)
    RETURNS int
    WITH SCHEMABINDING
AS
BEGIN
    DECLARE @part           nVarChar(20)
    DECLARE @Pos            Int
    DECLARE @LastPos        Int = 1
    DECLARE @MajorVersion   Int
    DECLARE @MinorVersion   Int
    DECLARE @BuildVersion   Int

    IF IsNull(@VersionString, N'') = N'' RETURN -1
    SET @Pos = CHARINDEX(N'.', @VersionString, @LastPos)
    IF IsNull(@Pos, 0) = 0             RETURN -1 -- no valid Version
    SET @Part = SUBSTRING(@VersionString, @LastPos, @Pos - 1)
    IF IsNumeric(@Part) = 1 SET @MajorVersion = CONVERT(int, @Part)
    SET @LastPos = @Pos + 1
    SET @Pos = CHARINDEX(N'.', @VersionString, @LastPos)
    SET @Part = SUBSTRING(@VersionString, @LastPos, @Pos - @LastPos)
    IF IsNumeric(@Part) = 1 SET @MinorVersion = CONVERT(int, @Part)
    SET @Part = SUBSTRING(@VersionString, @Pos + 1, 20)
    IF IsNumeric(@Part) = 1 SET @BuildVersion = CONVERT(int, @Part)

    RETURN IsNull(@MajorVersion, 0) * 10000 + IsNull(@MinorVersion, 0) * 100 + IsNull(@BuildVersion, 0)
END
GO
`
the numeric value is easy to compare in fn_CompareVersion:
'
CREATE FUNCTION {databaseOwner}[{objectQualifier}fn_CompareVersion]
(
    @Version        nVarChar(20),  -- Not Null
    @CurrentVersion nVarChar(20)   -- Not Null
)
    RETURNS int
    WITH SCHEMABINDING
AS
BEGIN
    DECLARE @VersionSerial  Int
    DECLARE @CurrentSerial  Int

    IF @CurrentVersion Is Null  -- Assembly Not Registered - Set ReturnCode = 0, so assembly is copied
        RETURN 0
    ELSE IF @Version = @CurrentVersion
        RETURN 2                -- Same Version - Set ReturnCode = 2, so assembly is only copied on repair
    ELSE BEGIN                  -- Different Versions, compare Major, Minor, Revision
        SET @VersionSerial = {databaseOwner}[{objectQualifier}VersionNumberSerial](@Version)
        SET @CurrentSerial = {databaseOwner}[{objectQualifier}VersionNumberSerial](@CurrentVersion)

        IF @VersionSerial > @CurrentSerial AND @CurrentSerial >= 0
            RETURN 1            -- Set ReturnCode = 1, so assembly is copied
        ELSE
            RETURN 3            -- Set ReturnCode = 3, so assembly is not copied
    END
    RETURN 3
END
GO
'

**Comment by galatrash-at-dnn** [comment]
@sleupold, please submit your suggested changes as a pull request so we can process it; otherwise, it is hard to apply your changes.

**Comment by sleupold** [comment]
@galatrash-at-dnn, thanks, will due over the weekend

**Comment by galatrash-at-dnn** [comment]
@sleupold any update on this issue?

**Pull request merged**